python/context.cpp

```diff
@@ -16,67 +18,158 @@
+    void set_gpu_id(pybind11::object gpu) {
+        gpu_id = py2optional<int>(gpu, "gpu id must be None, or a non-negative integer.", is_nonneg_int);
+    };
+
+    void set_num_threads(int threads) {
+        pyarb::assert_throw(is_nonneg_int(threads), "threads must be a non-negative integer.");
+        num_threads = threads;
+    };
+
+    const opt_int get_gpu_id()      const { return gpu_id; }
+    const int     get_num_threads() const { return num_threads; }
+    const bool    has_gpu()         const { return (gpu_id.value_or(-1) >= 0) ? true : false; }
+
+    // helper function to use arb::make_context(arb::proc_allocation)
+    arb::proc_allocation allocation() const {
+        return arb::proc_allocation(num_threads, gpu_id.value_or(-1));
+    }
+};
+
+// Helper template for printing C++ optional types in Python.
+// Prints either the value, or None if optional value is not set.
+template <typename T>
+std::string to_string(const arb::util::optional<T>& o) {
+    if (!o) return "None";
+
+    std::stringstream s;
+    s << *o;
+    return s.str();
+}
+
+std::string proc_alloc_string(const proc_allocation_shim& a) {
+    std::stringstream s;
+    s << "<hardware resource allocation: threads " << a.num_threads
+      << ", gpu id " << to_string(a.gpu_id);
+    s << ">";
+    return s.str();
+}
+
+// helper functions to wrap arb::make_context
+arb::context make_context_shim(const proc_allocation_shim& Alloc) {
+    return arb::make_context(Alloc.allocation());
+}
+
+template <typename Comm>
+arb::context make_context_shim(const proc_allocation_shim& Alloc, Comm comm) {
+    return arb::make_context(Alloc.allocation(), comm);
+}
+
 void register_contexts(pybind11::module& m) {
     using namespace std::string_literals;
     using namespace pybind11::literals;
+    using opt_int = arb::util::optional<int>;
+#ifdef ARB_MPI_ENABLED
+    using opt_mpi_comm = arb::util::optional<mpi_comm_shim>;
+#endif
 
-    pybind11::class_<arb::proc_allocation> proc_allocation(m, "proc_allocation");
+// proc_allocation
+    pybind11::class_<proc_allocation_shim> proc_allocation(m, "proc_allocation",
+        "Enumerates the computational resources on a node to be used for simulation.");
     proc_allocation
-        .def(pybind11::init<>())
-        .def(pybind11::init<int, int>(), "threads"_a, "gpu"_a=-1,
-             "Arguments:\n"
-             "  threads: The number of threads available locally for execution.\n"
-             "  gpu:     The index of the GPU to use, defaults to -1 for no GPU.\n")
-        .def_readwrite("threads", &arb::proc_allocation::num_threads,
+        .def(pybind11::init<int, pybind11::object>(),
+            "threads"_a=1, "gpu_id"_a=pybind11::none(),
+            "Construct an allocation with arguments:\n"
+            "  threads: The number of threads available locally for execution (defaults to 1).\n"
+            "  gpu_id:  The index of the GPU to use (defaults to None for no GPU).\n")
```

Comment:
It might be clearer to change the default text to `(default 1)` and `(default None)`.

Reply:
Done.